### ISSUE TITLE
Stackoverflow when setting up and calling object.Equals(object) on T : IEquatable<T>

### ISSUE THREAD

**Issue by BrunoJuchli** [moq/moq4#791]
### Version
Moq 4.10.1
.Net Framework 4.7.2

### Description
performing `mocked.Equals(new object()` throws a `StackOverflowException`. Conditions:

- mocked imlements `IEquatable<T>` interface
- setup for `object.Equals(object)` done which references the mock

### Repro

Given the following class `Foo` which implements the `IEquatable<Foo>` interface:

    public class Foo : IEquatable<Foo>
    {
        public virtual bool Equals(Foo other)
        {
            throw new NotImplementedException();
        }

        public override bool Equals(object obj)
        {
            throw new NotImplementedException();
        }

        public override int GetHashCode()
        {
            throw new NotImplementedException();
        }
    }

performing the following throws a `StackOverflowException`:

    var foo = new Mock<Foo>();

    foo.Setup(x => x.Equals((object)foo.Object)).Returns(true);

    foo.Object.Equals(new object());

### Workaround

setup `object.Equals(object)` like this instead:

    var foo = new Mock<Foo>();
    foo.Setup(x => x.Equals(It.Is<object>(o => ReferenceEquals(o, foo.Object)))).Returns(true);

**Comment by stakx** [comment]
Hi @BrunoJuchli, thanks for reporting. I'll take a closer look at this shortly to find out where Moq starts running in circles.

In theory, though, you shouldn't need such a setup at all:

> ```csharp
> foo.Setup(x => x.Equals((object)foo.Object)).Returns(true);
> ```

Because Moq should provide exactly that behavior out of the box:

https://github.com/moq/moq4/blob/efaeafd3860880c17b3ca16a8f5f5e2a882c84e6/src/Moq/Interception/InterceptionAspects.cs#L37-L43

**Comment by BrunoJuchli** [comment]
@stakx 
Thanks. Note: using [windbg you can catch `StackoverflowException`](https://stackoverflow.com/questions/4733878/how-to-debug-a-stackoverflowexception-in-net/49882734#49882734).

I'll look into the default behavior. I was under the impression it wasn't actually working like that, but I' was probably mistaken - I'll check tomorrow! :-)

**Comment by stakx** [comment]
@BrunoJuchli - you can reduce the repro code to the following three lines:

```csharp
var mock = new Mock<object>();
mock.Setup(m => m.Equals(mock.Object)).Returns(true);
mock.Object.Equals(new object());
```

**TL;DR:** Moq tries to match the argument expression in the setup against the concrete argument; it basically performs a `object.Equals(mock.Object, A)` (with `A` = `new object()`),  which results in a call to `mock.Object.Equals(A)`. Since there's a setup for `mock.Object.Equals`, this kicks off infinite recursion. IMO, this is a usage error on your part.

Some more detail:

1. Moq looks at the setup and sees an argument expression `mock.Object` which it evaluates to a constant value. Internally, it creates a suitable matcher (of type [`ConstantMatcher`](https://github.com/moq/moq4/blob/master/src/Moq/Matchers/ConstantMatcher.cs)) which will be tasked with matching concrete invocations' arguments against this value.

2. Along comes an interception, which Moq intercepts. It tries to match the provided argument `new object()` against the setup argument expression using the `ConstantMatcher` just mentioned.

3. `ConstantMatcher` is implemented simply in terms of `object.Equals`: It takes the concrete argument A := `new object()` and the constant derived from the setup argument expression C := `Mock.Object` and performs an equality check `object.Equals(C, A)`. This in turns calls `C.Equals(A)`, which triggers another interception that attempts to do the exact same thing all over again.

I think we should treat this as a usage error. Theoretically speaking, we could add logic to `ConstantMather` that tries to detect whether it might cause an infinite loop. This would be done by checking...

* whether C is a mock;
* whether C has a setup for the `object.Equals` method; and
* whether that setup has specified C as the first argument.

But this seems such a specific corner case that I think it's not worth the hassle.

Thoughts?

**Comment by stakx** [comment]
Oh, It turns out that a very trivial fix is possible. We can just switch the order in which we pass the two objects to `object.Equals`.

**Comment by BrunoJuchli** [comment]
@stakx 
I agree. Thanks for taking the time!

**Comment by BrunoJuchli** [comment]
@stakx 
I've now found out why I've done `foo.Setup(x => x.Equals((object)foo.Object)).Returns(true);`:

As soon as the object-to-mock overrides the `object Equals(object)` method the default setup doesn't apply anymore:

    public class Foo
    {
         public override bool Equals(object obj) => true;
    }

    var foo = new Mock<Foo>();
    foo.Object.Equals(foo.Object); // returns false

which kind-a makes sense to me: as long as you don't override `Equals` the mock behaves the same as the real thing (reference equality). And as soon as you specify something else, moq can't sensibly implement equality comparison and thus can have no sensible default.

How about: instead of returning `false` throwing an exception? If moq cannot deduce a sensible default behavior *and* a setup is missing, calling the Equals method is quite certainly something unexpected.
I know this behavior can be achieved by configuring the mock to be `strict`. We don't use - for other reasons - that style, however.

On the other hand, it seems to me that an `Equals` method returning `false` in case of reference equality is not something that is practical (i`m talking about production code, not mock behavior here) - and as such not something that should be of big importance.

Plus, only considering our own style of writing tests, it would be simplest if an mocked overriden equals method would just behave the same as an non-overriden one.
In tests, where we mock the object, we're not really interested in the comparison logic. Instead, if two objects are the same, we always just use one single object (mock), even if in reality they might come from different sources and be different instances. This, of course, opens a slight windows for bugs: the unit being tested could perform a `ReferenceEquals` instead of `object.Equals(object)` call. For now, we are fine with that.

**Comment by stakx** [comment]
> ```csharp
> foo.Object.Equals(foo.Object); // returns false
> ```
>
> [...] as long as you don't override Equals the mock behaves the same as the real thing (reference equality).

I am wondering why does that call returns `false`. Seems to me that it shouldn't do that. `object.ReferenceEquals(foo.Object, foo.Object) == true`, and the Moq's default implementation for `Equals` should follow suit.

> And as soon as you specify something else, moq can't sensibly implement equality comparison and thus can have no sensible default.

I suppose that by "specify something else", you mean (a) "override `Equals` in the mocked class", not (b) "setup the `Equals` method"?

Moq already deals with (b) by only providing the implementation for `Equals` if there is no setup for it.

It is currently possible to deal with (a) in two different ways:

* Make `Equals` non-interceptable:

   ```diff
    public class Foo
    {
   +     public sealed override bool Equals(object obj) => true;
   -     public override bool Equals(object obj) => true;
    }

* Provide a setup for `Equals` that calls the base implementation:

   ```diff
    var foo = new Mock<Foo>();
   +foo.Setup(f => f.Equals(It.IsAny<object>()).CallBase();
   +foo.Object.Equals(foo.Object); // returns true
   -foo.Object.Equals(foo.Object); // returns false
   ```

> How about: instead of returning false throwing an exception?

I don't think that would be sensible in the general case. Noone will expect `Equals` to throw (at least not on a loose mock). Which brings us back to my initial question of why `Equals` returns `false` in this particular case!?

**Comment by BrunoJuchli** [comment]
> I am wondering why that call returns `false` in the first place. It seems to me that it shouldn't do that: `object.ReferenceEquals(foo.Object, foo.Object) == true`, and Moq's default implementation for `Equals` should follow suit.

I suspect that `IsObjectMethod(invocation.Method)` is returning false because the overriding method is not declared on typeof(object)?:
> [moq4/src/Moq/Interception/InterceptionAspects.cs](https://github.com/moq/moq4/blob/efaeafd3860880c17b3ca16a8f5f5e2a882c84e6/src/Moq/Interception/InterceptionAspects.cs#L37-L43)
> Lines 37 to 43 in [efaeafd](/moq/moq4/commit/efaeafd3860880c17b3ca16a8f5f5e2a882c84e6)
> 
>  private static InterceptionAction HandleEquals(Invocation invocation, Mock mock) 
>  { 
>  	if (IsObjectMethod(invocation.Method) && !mock.Setups.Any(c => IsObjectMethod(c.Method, "Equals"))) 
>  	{ 
>  		invocation.Return(ReferenceEquals(invocation.Arguments.First(), mock.Object)); 
>  		return InterceptionAction.Stop; 
>  	}

--------------
> > And as soon as you specify something else, moq can't sensibly implement equality comparison and thus can have no sensible default.
> 
> I suppose that by "specify something else", you mean (a) "override `Equals` in the mocked class", not (b) "setup the `Equals` method"?

Yes, exactly.. I should have written "as soon as you override `Equals`".

> > How about: instead of returning false throwing an exception?
> 
> I don't think that would be sensible in the general case. Noone will expect `Equals` to throw (at least not on a loose mock). Which brings us back to my initial question of why `Equals` returns `false` in this particular case!?

I believe it would be quite strange, at least. On the other hand it would be safer because it would force the developer to consciously make a decision how object equality affects the test. Of course this is the same with all other methods, too. It's just that Equals & GetHashCode are often overlooked because "they are always there".

> Btw.:
> 
> > If moq cannot deduce a sensible default behavior [...]
> 
> It doesn't try very hard to begin with. All it checks are whether there's a setup for `Equals` or not. If not, Moq could additionally check (for mocked classes) whether the mocked type overrides `object.Equals` or implements `IEquatable<>`, and if so, do an automatic `.CallBase()`. If the mocked type does _not_ override that method or implement `IEquatable<>`, the current logic of doing an `object.ReferenceEquals` _is_ sensible as that's the default behavior of `Equals` on reference types.
> 
> Thoughts?

I agree that the current implementation for non-custom equals implementations absolutely makes sense. Changing it to always return `false` per default would just create a lot of unnecessary work.

I don't really think that it's necessarily a good idea to perform `.CallBase()` with custom equals implementations:

- could equals be based on internal state that is not available/settable if we mock the type?
- in "our" style of writing code, we try to separate data from state to some extent. Usually we only override equals on classes which are pure data containers. If I want to have the "real" implementation behavior I'd just better create the real thing than a mock. The thing is, we use mocks because we're not interested in all the data. Let me give you an example:

     public class Foo
     {
         public IReadOnlyList<Bar> SupportedBars { get; }
      }

    public class Bar : IEquatable<Bar>
    {
        public Guid Id { get; }
        public string Name { get; }
        public ComplexType ComplexProperty { get; }
    }

    public class FooBarMatcher
    {
        private IReadOnlyList<Foo> allFoos; // really, this is initialized by ctor of course..

        public IReadOnlyList<Foo> GetFoosSupporting(Bar bar)
        {
            return this.allFoos.Where(foo => foo.SupportedBars.Contains(bar)).ToList();
        }
    }

And now we have a test for `FooBarMatcher`. And here I'm mocking all the `Bar`s. I'm not really interested in the data a `Bar` has, just in equality. So I don't want to have `Equals` be configured with `CallBase()` because otherwise I have to setup a bunch of properties in order to cover the equality aspect.
It be much, much simpler though to just have `Equals` perform reference equality, because that's the simplest equality to use ;-)

Now, the reason I was bringing up "throwing an exception" per default is, because.. someone was writing tests for `FooBarMatcher` and got "crazy" because all the tests always failed while it worked in production. Of course there was no obvious cause. And you can only make it fail if you specifically write `foo.Equals((object)foo);` and not if you omit the cast to `object`, because we've set up that one already.
So it took quite a while to figure out what was the actual cause of the behavior.

Now, for our use case, ideally moq would treat overriden equals methods just like non-overriden ones (default behavior: reference equality check).
That's a breaking change however, and it might not be obvious. Hence I was thinking about throwing an exception, which is still a breaking change, but it would be an obvious one. Though personally I'd favor reference equality.

**Comment by stakx** [comment]
I'll study your example later when I have a little more time. For now:

> Though personally I'd favor reference equality.

As do I. :wink:

I think we should start by fixing the recognition logic for the `object.Equals` method.

https://github.com/moq/moq4/blob/efaeafd3860880c17b3ca16a8f5f5e2a882c84e6/src/Moq/Interception/InterceptionAspects.cs#L101

This check might be missing a call to `.GetBaseDefinition()`. It just might work correctly if changed to `method.GetBaseDefinition().DeclaringType == typeof(object)`, though that's currently just a guess.

**Comment by stakx** [comment]
@BrunoJuchli, I've opened a new issue (#802) regarding the misbehavior of `object.Equals`.

> > If moq cannot deduce a sensible default behavior [...]

> It doesn't try very hard to begin with. All it checks are whether there's a setup for `Equals` or not. If not, Moq could additionally check (for mocked classes) whether the mocked type overrides `object.Equals` or implements `IEquatable<>`, and if so, do an automatic `.CallBase()`.

I've found a unit test that underlines the above: If a mocked class overrides `object.Equals`, then Moq already should default to using that instead of overriding it with a reference equality check:

https://github.com/moq/moq4/blob/bf19ad64540b9e3e71b8a869b2b1ab204ff1dbcb/tests/Moq.Tests/MockFixture.cs#L399-L410

Still haven't studied your above example in detail, but I'll try to take it into consideration for #802.
